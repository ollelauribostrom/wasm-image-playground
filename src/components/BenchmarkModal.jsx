import React from 'react';
import Modal from 'react-modal';
import Icon from './Icon';
import BenchmarkTask from './BenchmarkTask';
import BenchmarkResult from './BenchmarkResult';
import Label from './Label';

const modalStyle = {
  content: {
    background: "transparent",
    border: 0
  },
  overlay: {
    background: "rgba(0, 0, 0, 0.9)",
    border: 0,
    zIndex: 2,            
  }
};

function BenchmarkModal({ isOpen, title, running, results, tasks, onStart, onRestart, onClose }) {
  return (
    <Modal isOpen={isOpen} style={modalStyle} >
      <div className="benchmark-content">
        <h1 className="benchmark-heading">Benchmark: {title}</h1>
        <Icon 
          name="restore"
          size="l"
          onClick={onRestart}
          className={`benchmark-reset-icon ${!running && results.length ? '' : 'hidden'}`}
        />
        <Icon 
          name="close"
          size="l"
          onClick={onClose}
          className="benchmark-close-icon"
        />
        <Label 
          text="Run Benchmark"
          size="large"
          onClick={onStart}
          className={`benchmark-button ${!running && !results.length ? '' : 'hidden'}`} 
        />
        {tasks.map(task => <BenchmarkTask {...task} />)}
        <div className={`benchmark-results ${results.length ? '' : 'hidden'}`}>
          {results.map(result => <BenchmarkResult {...result} />)}
        </div>
      </div>
    </Modal>
  )
}

export default BenchmarkModal;