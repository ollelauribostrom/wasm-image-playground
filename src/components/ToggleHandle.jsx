import React from 'react';
import { Icon } from './Icon';

type ToggleHandleProps = {
  onToggle: () => void
};

export function ToggleHandle({ onToggle }: ToggleHandleProps) {
  return (
    <div
      className="toggle__handle"
      onKeyPress={e => {
        const key = e.key || e.keyCode;
        if (key === 'Enter' || key === 13) {
          onToggle();
        }
      }}
      onClick={onToggle}
      role="button"
      tabIndex={0}
    >
      {<Icon name="toggle" />}
    </div>
  );
}
