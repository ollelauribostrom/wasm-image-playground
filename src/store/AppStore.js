import createStore from 'react-waterfall';
import ColorService from '../services/ColorService';

type AppState = {|
  count: number,
  color: string
|};

const initialState: AppState = {
  count: 0,
  color: '#fff'
};

const config = {
  initialState,
  actionsCreators: {
    click: async ({ count }: AppState): Promise<AppState> => ({
      count: count + 1,
      color: await ColorService.getRandomColor()
    })
  }
};

export const { Provider, connect, actions } = createStore(config);
