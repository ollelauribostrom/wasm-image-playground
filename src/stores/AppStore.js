import createStore from 'react-waterfall';
import ImageService from '../services/ImageService';
import type { Language, Filter } from '../services/types';

type AppState = {
  image?: Image,
  history: Array<Image>,
  status?: string,
  language: Language,
  isDragging: boolean,
  isLoading: boolean,
  showBenchmarkModal: boolean
};

export const initialState: AppState = {
  image: null,
  history: [],
  status: null,
  language: 'js',
  isDragging: false,
  isLoading: false,
  showBenchmarkModal: false
};

export const config = {
  initialState,
  actionsCreators: {
    setStatus: (state: AppState, actions: any, status: boolean): AppState => {
      return { status };
    },
    setLanguage: (state: AppState, actions: any, language: Language): AppState => {
      return { language };
    },
    setLoading: (state: AppState, actions: any, isLoading: boolean): AppState => {
      return { isLoading };
    },
    setDragging: (state: AppState, actions: any, isDragging: boolean): AppState => {
      return { isDragging };
    },
    toggleBenchmarkModal: ({ showBenchmarkModal }: AppState): AppState => {
      return { showBenchmarkModal: !showBenchmarkModal };
    },
    upload: async (state: AppState, actions: any, files: FileList): AppState => {
      const image = await ImageService.upload(files);
      return { image, isLoading: false };
    },
    applyFilter: async ({ image, history }: AppState, actions: any, filter: Filter): AppState => {
      const newImage = await ImageService.applyFilter(image, filter);
      const newHistory = [...history];
      newHistory.push(image);
      return { image: newImage, history: newHistory, isLoading: false };
    },
    runBenchmark: async (): AppState => {
      const benchmarkResults = await ImageService.benchmark();
      return { benchmarkResults, isLoading: false, showBenchmarkModal: true };
    },
    undo: ({ history }: AppState): AppState => {
      const newHistory = [...history];
      const newImage = newHistory.pop();
      return { image: newImage, history: newHistory, isLoading: false };
    },
    download: ({ image }: AppState): AppState => {
      ImageService.download(image);
      return { isLoading: false };
    },
    delete: (): AppState => {
      return { image: null, history: [] };
    }
  }
};

export const { Provider, connect, actions } = createStore(config);
