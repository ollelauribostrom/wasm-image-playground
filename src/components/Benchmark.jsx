import React, { Component } from 'react';
import { imageConverters } from 'imutils';
import BenchmarkModal from './BenchmarkModal';
import ImageService from '../services/ImageService';
import benchmarkImages from '../../benchmark/images.json'

class Benchmark extends Component {
  state = {
    taskIndex: 0,
    tasks: [],
    taskResults: {},
    results: [],
    error: null,
    running: false,
    showResult: false
  }

  componentWillMount() {
    ImageService.on('benchmarkUpdate', this.onUpdate);
    ImageService.on('benchmarkError', this.onError);
    ImageService.on('benchmarkComplete', this.onComplete);
  }

  componentWillUnmount() {
    ImageService.removeListener('benchmarkUpdate', this.onUpdate);
    ImageService.removeListener('benchmarkError', this.onError);
    ImageService.removeListener('benchmarkComplete', this.onComplete);
  }

  restart = async () => {
    if (!this.state.running && this.state.results) {
      await this.setState({ tasks: [], results: [], error: null, taskIndex: 0, taskResults: {} });
      this.start();
    }
  }

  start = () => {
    this.props.onStart();
    this.setState({ running: true });
    this.setup();
  }

  stop = () => {
    this.props.onStop();
  }

  setup = async () => {
    try {
      this.addOrUpdateTask({
        id: 'SETUP',
        info: 'Loading benchmark images',
        status: 'running',
        type: 'setup'
      });
      const images = await Promise.all(benchmarkImages.files.map(async image => {
        const res = await fetch(require(`../../benchmark/${image.path}`));
        const blob = await res.blob();
        const data = await imageConverters.toUint8ClampedArray(URL.createObjectURL(blob));
        return { data, containsFace: image.containsFace }
      }));
      await this.setState({ images });
      this.addOrUpdateTask({
        id: 'SETUP',
        info: `Loaded ${images.length} benchmark images`,
        status: 'done',
      });
      this.startTask(this.props.tasks[0]);
    } catch (error) {
      this.addOrUpdateTask({
        id: 'SETUP',
        info: 'Error when loading benchmark images',
        status: 'error',
      });
    }
  }

  startTask = task => {
    ImageService.postMessage({
      task,
      action: this.props.benchmarkType,
      images: this.state.images,
      results: task === 'getResults' ? this.state.taskResults : null
    });
  }

  addOrUpdateTask = task => {
    const tasks = [...this.state.tasks];
    const taskIndex = tasks.findIndex(t => t.id === task.id);
    if (taskIndex > -1) {
      tasks[taskIndex] = Object.assign(tasks[taskIndex], task);
    } else {
      tasks.push(task)
    }
    this.setState({ tasks });

    if (task.status === 'done' && task.id !== 'SETUP') {
      const taskResults = Object.assign({}, this.state.taskResults);
      const currentTask = this.props.tasks[this.state.taskIndex];
      const nextTask = this.props.tasks[this.state.taskIndex + 1];
      taskResults[currentTask] = task.result;
      this.setState({ taskResults, taskIndex: this.state.taskIndex + 1 });
      if (nextTask) {
        this.startTask(nextTask);
      }
    }
  }

  onUpdate = ({ data }) => {
    this.addOrUpdateTask(data.task);
  }
  
  onError = ({ data }) => {
    this.setState({
      error: data.error,
      running: false
    });
    this.props.onStop();
  }

  onComplete = ({ data }) => {
    this.setState({ running: false, results: data.results });
    this.props.onStop();
  }

  onClose = () => {
    this.setState({
      isOpen: false
    });
  }

  render() {
    return (
      <BenchmarkModal
        isOpen={this.props.isOpen}
        running={this.state.running}
        tasks={this.state.tasks}
        results={this.state.results}
        title={this.props.title}
        onStart={this.start}
        onRestart={this.restart}
        onClose={this.props.onClose}
      />
    )
  }
}

export default Benchmark;
