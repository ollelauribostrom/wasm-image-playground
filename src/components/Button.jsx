import React from 'react';
import { Icon } from './Icon';
import { ToggleHandle } from './ToggleHandle';

type ButtonProps = {
  icon?: Icon,
  label?: string,
  title?: string,
  isDisabled?: boolean,
  onClick?: () => void,
  onToggle?: () => void,
  customClassName?: string,
  href?: string,
  target?: string,
  rel?: string
};

export function Button({
  icon,
  label,
  title,
  isDisabled,
  onClick,
  onToggle,
  customClassName,
  href,
  target,
  rel
}: ButtonProps) {
  let className = `button ${customClassName}`;
  if (isDisabled) {
    className += 'button--disabled';
  }
  if (href && !isDisabled) {
    return (
      <a className={className} href={href} target={target} rel={rel} title={title}>
        {label}
        {icon}
      </a>
    );
  }
  return (
    <div
      onKeyPress={e => {
        const key = e.key || e.keyCode;
        if (key === 'Enter' || key === 13) {
          if (!isDisabled) {
            onClick();
          }
        }
      }}
      onClick={e => !isDisabled && onClick(e)}
      className={`button ${customClassName}`}
      type="button"
      title={title}
      role="button"
      tabIndex={0}
    >
      {label ? <span>{label}</span> : null}
      {icon}
      {onToggle ? (
        <ToggleHandle
          onToggle={e => {
            e.stopPropagation();
            onToggle();
          }}
        />
      ) : null}
    </div>
  );
}

Button.defaultProps = {
  icon: null,
  label: null,
  title: '',
  isDisabled: false,
  onClick: () => {},
  onToggle: null,
  customClassName: '',
  href: null,
  target: null,
  rel: null
};
