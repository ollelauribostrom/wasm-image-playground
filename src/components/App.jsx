import React from 'react';
import { Provider } from '../store/AppStore';
import ErrorBoundary from './ErrorBoundary';
import Reactor from './Reactor';
import Count from './Count';
import '../styles/main.css';

export default function App() {
  return (
    <Provider>
      <ErrorBoundary>
        <Reactor />
        <Count />
      </ErrorBoundary>
    </Provider>
  );
}
