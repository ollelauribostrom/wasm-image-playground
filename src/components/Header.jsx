import React from 'react';
import Icon from './Icon';
import Button from './Button';
import { connect, actions } from '../stores/AppStore';
import { Filters, Languages } from '../services/types';
import type { Language } from '../services/types';

type HeaderProps = {
  image: Image,
  language: Language,
  history: Array<Image>,
  isLoading: boolean
};

function Header({ image, language, history, isLoading }: HeaderProps) {
  return (
    <div className="header">
      <div className="header__logo">
        <Icon name="image" />
      </div>
      <div className="header__toolbar">
        <Button
          customClassName="toolbar__button toolbar__button--filter"
          title="Blur image"
          icon={<Icon name="blur" />}
          onClick={async () => {
            actions.setLoading(true);
            await actions.applyFilter(Filters.BoxBlur);
          }}
          isDisabled={isLoading}
        />
        <Button
          customClassName="toolbar__button toolbar__button--filter"
          title="Convert image to grayscale"
          icon={<Icon name="bw" />}
          onClick={async () => {
            actions.setLoading(true);
            await actions.applyFilter(Filters.Grayscale);
          }}
          isDisabled={isLoading}
        />
        <Button
          customClassName="toolbar__button toolbar__button--filter"
          title="Sharpen image"
          icon={<Icon name="sharpen" />}
          onClick={async () => {
            actions.setLoading(true);
            await actions.applyFilter(Filters.Sharpen);
          }}
          isDisabled={isLoading}
        />
        <Button
          customClassName="toolbar__button toolbar__button--filter"
          title="Invert image"
          icon={<Icon name="invert" />}
          onClick={async () => {
            actions.setLoading(true);
            await actions.applyFilter(Filters.Invert);
          }}
          isDisabled={isLoading}
        />
        <Button
          customClassName="toolbar__button toolbar__button--filter"
          title="Apply cooling filter"
          icon={<Icon name="cool" />}
          onClick={async () => {
            actions.setLoading(true);
            await actions.applyFilter(Filters.Cooling);
          }}
          isDisabled={isLoading}
        />
        <Button
          customClassName="toolbar__button toolbar__button--benchmark"
          title="Open benchmark "
          icon={<Icon name="benchmark" />}
          onClick={() => {
            actions.toggleBenchmarkModal();
          }}
        />
        <Button
          customClassName="toolbar__button toolbar__button--action"
          title="Undo last action"
          icon={<Icon name="undo" />}
          onClick={() => {
            actions.setLoading(true);
            actions.undo();
          }}
          isDisabled={isLoading || !history.length}
        />
        <Button
          customClassName="toolbar__button toolbar__button--action"
          title="Download"
          icon={<Icon name="download" />}
          onClick={() => {
            actions.setLoading(true);
            actions.download();
          }}
          isDisabled={isLoading || !image}
        />
        <Button
          customClassName="toolbar__button toolbar__button--action"
          title="Delete"
          icon={<Icon name="delete" />}
          onClick={() => {
            actions.setLoading(true);
            actions.delete();
          }}
          isDisabled={isLoading}
        />
        <Button
          customClassName="toolbar__button toolbar__button--action"
          title="Current language mode"
          icon={<Icon name={language} />}
          onToggle={() =>
            actions.setLanguage(
              language === Languages.JavaScript ? Languages.WebAssembly : Languages.JavaScript
            )
          }
          isDisabled={isLoading}
        />
      </div>
    </div>
  );
}

export default connect(state => state)(Header);
