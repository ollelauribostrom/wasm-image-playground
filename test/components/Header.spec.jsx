import React from 'react';
import { shallow, mount } from 'enzyme';
import Header from '../../src/components/Header';
import { actions } from '../../src/store/AppStore';
import { Button } from '../../src/components/Button';
import { ToggleHandle } from '../../src/components/ToggleHandle';

jest.mock('../../src/store/AppStore', () => ({
  connect: () => arg => arg,
  actions: {
    setStatus: jest.fn(),
    setLoading: jest.fn(),
    setLanguage: jest.fn(),
    toggleBenchmarkModal: jest.fn(),
    applyFilter: jest.fn(),
    runBenchmark: jest.fn(),
    undo: jest.fn(),
    download: jest.fn(),
    delete: jest.fn()
  }
}));

const ButtonIndex = {
  Blur: 0,
  Grayscale: 1,
  Sharpen: 2,
  Invert: 3,
  Cool: 4,
  Benchmark: 5,
  Undo: 6,
  Download: 7,
  Delete: 8
};

describe('Tests for Header', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Header language="js" />);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders correctly when language is wasm', () => {
    const wrapper = shallow(<Header language="wasm" />);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders correctly when app is loading', () => {
    const wrapper = shallow(<Header language="js" isLoading />);
    expect(wrapper).toMatchSnapshot();
  });
  it('handles click on the blur button', () => {
    const wrapper = shallow(<Header language="js" />);
    wrapper
      .find(Button)
      .at(ButtonIndex.Blur)
      .simulate('click');
    expect(actions.setLoading).toHaveBeenCalledWith(true);
    expect(actions.applyFilter).toHaveBeenCalledWith('blur');
  });
  it('handles click on the grayscale button', () => {
    const wrapper = shallow(<Header language="js" />);
    wrapper
      .find(Button)
      .at(ButtonIndex.Grayscale)
      .simulate('click');
    expect(actions.setLoading).toHaveBeenCalledWith(true);
    expect(actions.applyFilter).toHaveBeenCalledWith('grayscale');
  });
  it('handles click on the sharpen button', () => {
    const wrapper = shallow(<Header language="js" />);
    wrapper
      .find(Button)
      .at(ButtonIndex.Sharpen)
      .simulate('click');
    expect(actions.setLoading).toHaveBeenCalledWith(true);
    expect(actions.applyFilter).toHaveBeenCalledWith('sharpen');
  });
  it('handles click on the invert button', () => {
    const wrapper = shallow(<Header language="js" />);
    wrapper
      .find(Button)
      .at(ButtonIndex.Invert)
      .simulate('click');
    expect(actions.setLoading).toHaveBeenCalledWith(true);
    expect(actions.applyFilter).toHaveBeenCalledWith('invert');
  });
  it('handles click on the cool button', () => {
    const wrapper = shallow(<Header language="js" />);
    wrapper
      .find(Button)
      .at(ButtonIndex.Cool)
      .simulate('click');
    expect(actions.setLoading).toHaveBeenCalledWith(true);
    expect(actions.applyFilter).toHaveBeenCalledWith('cool');
  });
  it('handles click on the benchmark button', () => {
    const wrapper = shallow(<Header language="js" />);
    wrapper
      .find(Button)
      .at(ButtonIndex.Benchmark)
      .simulate('click');
    expect(actions.toggleBenchmarkModal).toHaveBeenCalled();
  });
  it('handles click on the undo button', () => {
    const wrapper = shallow(<Header language="js" />);
    wrapper
      .find(Button)
      .at(ButtonIndex.Undo)
      .simulate('click');
    expect(actions.setLoading).toHaveBeenCalledWith(true);
    expect(actions.undo).toHaveBeenCalled();
  });
  it('handles click on the download button', () => {
    const wrapper = shallow(<Header language="js" />);
    wrapper
      .find(Button)
      .at(ButtonIndex.Download)
      .simulate('click');
    expect(actions.setLoading).toHaveBeenCalledWith(true);
    expect(actions.download).toHaveBeenCalled();
  });
  it('handles click on the delete button', () => {
    const wrapper = shallow(<Header language="js" />);
    wrapper
      .find(Button)
      .at(ButtonIndex.Delete)
      .simulate('click');
    expect(actions.setLoading).toHaveBeenCalledWith(true);
    expect(actions.delete).toHaveBeenCalled();
  });
  it('handles click on the language toggle handle (js => wasm)', () => {
    const wrapper = mount(<Header language="js" />);
    wrapper.find(ToggleHandle).simulate('click');
    expect(actions.setLanguage).toHaveBeenCalledWith('wasm');
    wrapper.unmount();
  });
  it('handles click on the language toggle handle (wasm => js)', () => {
    const wrapper = mount(<Header language="wasm" />);
    wrapper.find(ToggleHandle).simulate('click');
    expect(actions.setLanguage).toHaveBeenCalledWith('js');
    wrapper.unmount();
  });
});
