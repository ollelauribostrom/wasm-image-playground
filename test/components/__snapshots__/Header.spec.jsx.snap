// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for Header renders correctly 1`] = `
<div
  className="header"
>
  <div
    className="header__logo"
  >
    <Icon
      name="image"
    />
  </div>
  <div
    className="header__toolbar"
  >
    <Button
      customClassName="toolbar__button toolbar__button--filter"
      href={null}
      icon={
        <Icon
          name="blur"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Blur image"
    />
    <Button
      customClassName="toolbar__button toolbar__button--filter"
      href={null}
      icon={
        <Icon
          name="bw"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Convert image to grayscale"
    />
    <Button
      customClassName="toolbar__button toolbar__button--filter"
      href={null}
      icon={
        <Icon
          name="sharpen"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Sharpen image"
    />
    <Button
      customClassName="toolbar__button toolbar__button--filter"
      href={null}
      icon={
        <Icon
          name="invert"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Invert image"
    />
    <Button
      customClassName="toolbar__button toolbar__button--filter"
      href={null}
      icon={
        <Icon
          name="cool"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Apply cooling filter"
    />
    <Button
      customClassName="toolbar__button toolbar__button--benchmark"
      href={null}
      icon={
        <Icon
          name="benchmark"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Open benchmark "
    />
    <Button
      customClassName="toolbar__button toolbar__button--action"
      href={null}
      icon={
        <Icon
          name="undo"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Undo last action"
    />
    <Button
      customClassName="toolbar__button toolbar__button--action"
      href={null}
      icon={
        <Icon
          name="download"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Download"
    />
    <Button
      customClassName="toolbar__button toolbar__button--action"
      href={null}
      icon={
        <Icon
          name="delete"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Delete"
    />
    <Button
      customClassName="toolbar__button toolbar__button--action"
      href={null}
      icon={
        <Icon
          name="js"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={[Function]}
      rel={null}
      target={null}
      title="Current language mode"
    />
  </div>
</div>
`;

exports[`Tests for Header renders correctly when app is loading 1`] = `
<div
  className="header"
>
  <div
    className="header__logo"
  >
    <Icon
      name="image"
    />
  </div>
  <div
    className="header__toolbar"
  >
    <Button
      customClassName="toolbar__button toolbar__button--filter"
      href={null}
      icon={
        <Icon
          name="blur"
        />
      }
      isDisabled={true}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Blur image"
    />
    <Button
      customClassName="toolbar__button toolbar__button--filter"
      href={null}
      icon={
        <Icon
          name="bw"
        />
      }
      isDisabled={true}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Convert image to grayscale"
    />
    <Button
      customClassName="toolbar__button toolbar__button--filter"
      href={null}
      icon={
        <Icon
          name="sharpen"
        />
      }
      isDisabled={true}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Sharpen image"
    />
    <Button
      customClassName="toolbar__button toolbar__button--filter"
      href={null}
      icon={
        <Icon
          name="invert"
        />
      }
      isDisabled={true}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Invert image"
    />
    <Button
      customClassName="toolbar__button toolbar__button--filter"
      href={null}
      icon={
        <Icon
          name="cool"
        />
      }
      isDisabled={true}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Apply cooling filter"
    />
    <Button
      customClassName="toolbar__button toolbar__button--benchmark"
      href={null}
      icon={
        <Icon
          name="benchmark"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Open benchmark "
    />
    <Button
      customClassName="toolbar__button toolbar__button--action"
      href={null}
      icon={
        <Icon
          name="undo"
        />
      }
      isDisabled={true}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Undo last action"
    />
    <Button
      customClassName="toolbar__button toolbar__button--action"
      href={null}
      icon={
        <Icon
          name="download"
        />
      }
      isDisabled={true}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Download"
    />
    <Button
      customClassName="toolbar__button toolbar__button--action"
      href={null}
      icon={
        <Icon
          name="delete"
        />
      }
      isDisabled={true}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Delete"
    />
    <Button
      customClassName="toolbar__button toolbar__button--action"
      href={null}
      icon={
        <Icon
          name="js"
        />
      }
      isDisabled={true}
      label={null}
      onClick={[Function]}
      onToggle={[Function]}
      rel={null}
      target={null}
      title="Current language mode"
    />
  </div>
</div>
`;

exports[`Tests for Header renders correctly when language is wasm 1`] = `
<div
  className="header"
>
  <div
    className="header__logo"
  >
    <Icon
      name="image"
    />
  </div>
  <div
    className="header__toolbar"
  >
    <Button
      customClassName="toolbar__button toolbar__button--filter"
      href={null}
      icon={
        <Icon
          name="blur"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Blur image"
    />
    <Button
      customClassName="toolbar__button toolbar__button--filter"
      href={null}
      icon={
        <Icon
          name="bw"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Convert image to grayscale"
    />
    <Button
      customClassName="toolbar__button toolbar__button--filter"
      href={null}
      icon={
        <Icon
          name="sharpen"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Sharpen image"
    />
    <Button
      customClassName="toolbar__button toolbar__button--filter"
      href={null}
      icon={
        <Icon
          name="invert"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Invert image"
    />
    <Button
      customClassName="toolbar__button toolbar__button--filter"
      href={null}
      icon={
        <Icon
          name="cool"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Apply cooling filter"
    />
    <Button
      customClassName="toolbar__button toolbar__button--benchmark"
      href={null}
      icon={
        <Icon
          name="benchmark"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Open benchmark "
    />
    <Button
      customClassName="toolbar__button toolbar__button--action"
      href={null}
      icon={
        <Icon
          name="undo"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Undo last action"
    />
    <Button
      customClassName="toolbar__button toolbar__button--action"
      href={null}
      icon={
        <Icon
          name="download"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Download"
    />
    <Button
      customClassName="toolbar__button toolbar__button--action"
      href={null}
      icon={
        <Icon
          name="delete"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={null}
      rel={null}
      target={null}
      title="Delete"
    />
    <Button
      customClassName="toolbar__button toolbar__button--action"
      href={null}
      icon={
        <Icon
          name="wasm"
        />
      }
      isDisabled={false}
      label={null}
      onClick={[Function]}
      onToggle={[Function]}
      rel={null}
      target={null}
      title="Current language mode"
    />
  </div>
</div>
`;
